arches/app/src/components/ControlledListManager/editor/ImageMetadata.vue

```diff
@@ -0,0 +1,424 @@
+<script setup lang="ts">
+import arches from "arches";
+import { computed, ref, inject } from "vue";
+import { useGettext } from "vue3-gettext";
+
+import Button from "primevue/button";
+import Column from "primevue/column";
+import Dropdown from "primevue/dropdown";
+import DataTable from "primevue/datatable";
+import InputText from "primevue/inputtext";
+import { useToast } from "primevue/usetoast";
+
+import {
+    deleteImage,
+    deleteMetadata,
+    upsertMetadata,
+} from "@/components/ControlledListManager/api.ts";
+import {
+    DANGER,
+    DEFAULT_ERROR_TOAST_LIFE,
+    ERROR,
+    METADATA_CHOICES,
+    itemKey,
+} from "@/components/ControlledListManager/constants.ts";
+import { languageNameFromCode } from "@/components/ControlledListManager/utils.ts";
+import AddMetadata from "@/components/ControlledListManager/editor/AddMetadata.vue";
+
+import type { Ref } from "vue";
+import type { DataTableRowEditInitEvent } from "primevue/datatable";
+import type {
+    ControlledListItem,
+    ControlledListItemImage,
+    ControlledListItemImageMetadata,
+    LabeledChoice,
+    NewOrExistingControlledListItemImageMetadata,
+} from "@/types/ControlledListManager";
+
+const toast = useToast();
+const { $gettext } = useGettext();
+
+const metadataTypeHeader = $gettext("Metadata type");
+const metadataValueHeader = $gettext("Value");
+const languageHeader = $gettext("Language");
+
+const item = inject(itemKey) as Ref<ControlledListItem>;
+const { image } = defineProps<{ image: ControlledListItemImage }>();
+const editingRows = ref<NewOrExistingControlledListItemImageMetadata[]>([]);
+const rowIndexToFocus = ref(-1);
+const editorRef: Ref<HTMLDivElement | null> = ref(null);
+
+const labeledChoices: LabeledChoice[] = [
+    {
+        type: METADATA_CHOICES.title,
+        label: $gettext("Title"),
+    },
+    {
+        type: METADATA_CHOICES.alternativeText,
+        label: $gettext("Alternative text"),
+    },
+    {
+        type: METADATA_CHOICES.description,
+        label: $gettext("Description"),
+    },
+    {
+        type: METADATA_CHOICES.attribution,
+        label: $gettext("Attribution"),
+    },
+];
+
+const metadataLabel = (metadataType: string) => {
+    return labeledChoices.find((choice) => choice.type === metadataType)!.label;
+};
+
+const saveMetadata = async (event: DataTableRowEditInitEvent) => {
+    // normalize new metadata numbers to null
+    const normalizedNewData: ControlledListItemImageMetadata = {
+        ...event.newData,
+        id: typeof event.newData.id === "string" ? event.newData.id : null,
+    };
+    try {
+        const upsertedMetadata: ControlledListItemImageMetadata =
+            await upsertMetadata(normalizedNewData);
+        if (normalizedNewData.id) {
+            updateImageMetadata(upsertedMetadata);
+        } else {
+            appendImageMetadata(upsertedMetadata);
+            removeImageMetadata(event.newData);
+        }
+    } catch (error) {
```

Comment:
Yeah that's fair, I think I ended up doing this to avoid having to `let` the result of the request outside the `try`, but it does have the downside you describe, will update 👍